libs/full/collectives/tests/unit/communication_set.cpp

```diff
@@ -86,12 +86,13 @@ namespace hpx { namespace traits {
         }
 
         template <typename Result>
-        Result get(std::size_t, test::get_connected_to_zero)
+        Result get(std::size_t which, test::get_connected_to_zero)
         {
             // first forward request to parent
-            if (communicator_.connect_to_ != communicator_.site_)
+            if (communicator_.connect_to_ != communicator_.site_ &&
+                which == communicator_.site_)
             {
-                hpx::sync(test::get_connected_to_zero_action{},
+                hpx::apply(test::get_connected_to_zero_action{},
```

Comment:
Does this synchronize with the code after this through the mutex? I guess this is somehow the main trigger to potential deadlocks?

Reply:
The test does not depend on the synchronization. And yes, this was possibly causing deadlocks.